R/rows.R

```diff
@@ -65,21 +74,36 @@ by_slice <- function(.d, ..f, ..., .labels = TRUE) {
   }
 
   if (!dplyr::is.grouped_df(.d)) {
-    ..f(.d, ...)
-  } else {
-    dots <- substitute(pairlist(...))
-    calling_env <- parent.frame()
-    by_slice_impl(.d, ..f, dots, .labels, calling_env)
+    stop(".d must be a sliced data frame", call. = FALSE)
+  }
+  if (!.collate %in% c("list", "rows", "cols")) {
+    stop(".collate should be `list`, `rows` or `cols`", call. = FALSE)
   }
+
+  set_sliced_env(.d, .labels, .collate, .to, environment(), ".d")
+  .Call("by_slice_impl", environment(), ".d", "..f")
```

Comment:
You should set the `PACKAGE` argument here; otherwise R is forced to lookup the symbol by scanning through all loaded libraries (and this can take a non-trivial amount of time if many libraries are loaded).